agent/app/agent.go

```diff
@@ -0,0 +1,404 @@
+)
+
+const (
+	containerChangeEventStreamName             = "ContainerChange"
+	deregisterContainerInstanceEventStreamName = "DeregisterContainerInstance"
+	clusterMismatchErrorFormat                 = "Data mismatch; saved cluster '%v' does not match configured cluster '%v'. Perhaps you want to delete the configured checkpoint file?"
+	instanceIDMismatchErrorFormat              = "Data mismatch; saved InstanceID '%s' does not match current InstanceID '%s'. Overwriting old datafile"
+	instanceTypeMismatchErrorFormat            = "The current instance type does not match the registered instance type. Please revert the instance type change, or alternatively launch a new instance: %v"
+)
+
+// agent interface is used by the app runner to interact with the ecsAgent
+// object. Its purpose is to mostly demonstrate how to interact with the
+// ecsAgent type.
+type agent interface {
+	// printVersion prints the Agent version string
+	printVersion() int
+	// start starts the Agent execution
+	start() int
+}
+
+// ecsAgent wraps all the entities needed to start the ECS Agent execution.
+// after creating it via
+// the newAgent() method
+type ecsAgent struct {
+	ctx                  context.Context
+	ec2MetadataClient    ec2.EC2MetadataClient
+	cfg                  *config.Config
+	dockerClient         engine.DockerClient
+	containerInstanceArn string
+	credentialProvider   *aws_credentials.Credentials
+}
+
+// newAgent returns a new ecsAgent object
+func newAgent(
+	ctx context.Context,
+	blackholeEC2Metadata bool,
+	acceptInsecureCert *bool) (agent, error) {
+
+	ec2MetadataClient := ec2.NewEC2MetadataClient(nil)
+	if blackholeEC2Metadata {
+		ec2MetadataClient = ec2.NewBlackholeEC2MetadataClient()
+	}
+
+	log.Info("Loading configuration")
+	cfg, err := config.NewConfig(ec2MetadataClient)
+	if err != nil {
+		// All required config values can be inferred from EC2 Metadata,
+		// so this error could be transient.
+		log.Criticalf("Error loading config: %v", err)
+		return nil, err
+	}
+	cfg.AcceptInsecureCert = aws.BoolValue(acceptInsecureCert)
+	if cfg.AcceptInsecureCert {
+		log.Warn("SSL certificate verification disabled. This is not recommended.")
+	}
+	log.Debugf("Loaded config: %s", cfg.String())
+
+	dockerClient, err := engine.NewDockerGoClient(dockerclient.NewFactory(cfg.DockerEndpoint), cfg)
+	if err != nil {
+		// This is also non terminal in the current config
+		log.Criticalf("Error creating Docker client: %v", err)
+		return nil, err
+	}
+
+	return &ecsAgent{
+		ctx:               ctx,
+		ec2MetadataClient: ec2MetadataClient,
+		cfg:               cfg,
+		dockerClient:      dockerClient,
+		// We instantiate our own credentialProvider for use in acs/tcs. This tries
+		// to mimic roughly the way it's instantiated by the SDK for a default
+		// session.
+		credentialProvider: defaults.CredChain(defaults.Config(), defaults.Handlers()),
+	}, nil
+
+}
+
+func (agent *ecsAgent) printVersion() int {
+	version.PrintVersion(agent.dockerClient)
```

Comment:
I don't understand this/these abstraction(s).

`agent.printVersion()`  returns the agent version. Why does it need the dockerclient to figure out which version of agent is running?